pkg/migrations/coordinator.go

```diff
@@ -0,0 +1,61 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package migrations
+
+import (
+	"context"
+	"fmt"
+	"slices"
+)
+
+// Coordinator is responsible for executing a series of database actions in a specific order.
+// It ensures that each action is executed only once, even if it is added multiple times.
+type Coordinator struct {
+	actions map[string]DBAction
+	order   []string
+}
+
+func NewCoordinator(actions []DBAction) *Coordinator {
+	actionsMap := make(map[string]DBAction, len(actions))
+	order := make([]string, 0)
+	for _, action := range actions {
+		if act, exists := actionsMap[action.ID()]; exists {
+			order = moveIdxToLast(order, slices.Index(order, act.ID()))
+		} else {
+			actionsMap[action.ID()] = action
+			order = append(order, action.ID())
+		}
+	}
+	return &Coordinator{
+		actions: actionsMap,
+		order:   order,
+	}
+}
+
+// Execute runs all actions in the order they were added to the coordinator.
+func (c *Coordinator) Execute(ctx context.Context) error {
+	for _, id := range c.order {
+		action, exists := c.actions[id]
+		if !exists {
+			return fmt.Errorf("action %s not found", id)
+		}
+		if err := action.Execute(ctx); err != nil {
+			return fmt.Errorf("failed to execute action %s: %w", id, err)
+		}
+	}
+	return nil
+}
+
+func moveIdxToLast(actions []string, idx int) []string {
+	if idx < 0 || idx >= len(actions) {
+		return actions
+	}
+	duplicate := actions[idx]
+	actions = append(actions[:idx], actions[idx+1:]...)
+	if len(actions) > 0 && actions[len(actions)-1] == duplicate {
+		// If the last action is the same as the one we want to move, no need to change order
+		return actions
+	}
```

Comment:
All test cases still pass when I remove this check. Missing coverage or redundant check?

AIUI this tries to ensure that any duplicate actions occur at or near the end of the actions list; I think iterating over the slice first to last and simply removing duplicates gets us to the same result?

Reply:
An alternative design might be to have the coordinator apply rules, in order, to the actions list.

The only rule we'd need at this stage is a `removeDuplicateActions` rule AIUI.